### DISCUSSION TITLE
A few questions about field inversion using my own mesh

### DISCUSSION THREAD

**Discussion by SomeChilliness** [mdolab/dafoam#718] [General]
Hello all,

I am a new dafoam user and I am trying to do the field inversion for a NASA bump 2D case(SA model, same BCs and mesh but changing "empty" to "symmetry"). 
[blockMeshDict.txt](https://github.com/user-attachments/files/18081925/blockMeshDict.txt)

I get the runScript.py by modifying the one in the PeriodicHill_FieldInversion tutorial.
([runScript.txt](https://github.com/user-attachments/files/18081713/runScript.txt) Please find the whole file here)
And the modifications are:
1. using my own Input Parameters: 
```
    U0 = 69.44  
    nuTilda0 = 6.93e-5  
    J0 = 31220.744860210347 
    nCells = 3526
```

2. removing fvSource because I think it is not needed in this case

3. using this for "FI" in "objFunc" because I would like to use the whole velocity field in "UData" for field inversion:
```
            "UData": {
                "type": "fieldInversion",
                "source": "boxToCell",
                "min": [59.44, -10.0, -10.0],
                "max": [79.44, 10.0, 10.0],
                "data": "UData",
                "scale": 1,
                "addToAdjoint": True,
                "weightedSum": True,
                "weight": 1 / J0,
            },
```

4. adding this so that checkMesh could pass:
```
   "checkMeshThreshold": {
    "maxAspectRatio": 165000, #1000.0,
    "maxNonOrth": 70.0,
    "maxSkewness": 4.0,
    "maxIncorrectlyOrientedFaces": 0,
    },
```

Then I ran it but did not acquire results as expected:

1. _the skin friction coefficient in my postprocess significantly increased comparing with results using SA model_ 
2. _the velocity field of last iter did not get close to the one in "UData", and in some position, the difference could be quite large_
3. _then I checked the betaFieldInversion of last iter, it looks weird: it seems to be the same value everywhere, which is not the same story as PeriodicHill_FieldInversion tutorial case_ 
[betaFieldInversion.txt](https://github.com/user-attachments/files/18082438/betaFieldInversion.txt)
4. _when checking the logs, I found that IPOPT did not take many iters, and I got some warnings:_
```
Calculating [dRdW]^T * Psi using reverse-mode AD
Solving Linear Equation... 28 s
Main iteration 0 KSP Residual norm 0.000000000000e+00 28 s
Main iteration 0 KSP Residual norm 0.000000000000e+00 28 s 
Total iterations 0
Solving Linear Equation... Completed! 28 s
Residual tolerance satisfied, solution finished!
Calculating [dRdField]^T * Psi using reverse-mode AD
/home/l/dafoam/packages/miniconda3/lib/python3.8/site-packages/openmdao/core/component.py:140: OMDeprecationWarning:'cruise.coupling.solver' <class DAFoamSolver>: It appears this component mixes distributed/non-distributed inputs and outputs, so it may break starting with OpenMDAO 3.25, where the convention used when passing data between distributed and non-distributed inputs and outputs within a matrix free component will change. See https://github.com/OpenMDAO/POEMs/blob/master/POEM_075.md for details.
/home/l/dafoam/packages/miniconda3/lib/python3.8/site-packages/openmdao/core/total_jac.py:1636: DerivativesWarning:Constraints or objectives [('cruise.aero_post.functionals.FI', inds=[0])] cannot be impacted by the design variables of the problem.
/home/l/dafoam/packages/miniconda3/lib/python3.8/site-packages/openmdao/core/total_jac.py:1665: DerivativesWarning:Design variables [('dvs.beta', inds=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 298, 299, 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 322, 323, 324, 325, 326, 327, 328, 329, 330, 331, 332, 333, 334, 335, 336, 337, 338, 339, 340, 341, 342, 343, 344, 345, 346, 347, 348, 349, 350, 351, 352, 353, 354, 355, 356, 357, 358, 359, 360, 361, 362, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411, 412, 413, 414, 415, 416, 417, 418, 419, 420, 421, 422, 423, 424, 425, 426, 427, 428, 429, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 440, 441, 442, 443, 444, 445, 446, 447, 448, 449, 450, 451, 452, 453, 454, 455, 456, 457, 458, 459, 460, 461, 462, 463, 464, 465, 466, 467, 468, 469, 470, 471, 472, 473, 474, 475, 476, 477, 478, 479, 480, 481, 482, 483, 484, 485, 486, 487, 488, 489, 490, 491, 492, 493, 494, 495, 496, 497, 498, 499, 500, 501, 502, 503, 504, 505, 506, 507, 508, 509, 510, 511, 512, 513, 514, 515, 516, 517, 518, 519, 520, 521, 522, 523, 524, 525, 526, 527, 528, 529, 530, 531, 532, 533, 534, 535, 536, 537, 538, 539, 540, 541, 542, 543, 544, 545, 546, 547, 548, 549, 550, 551, 552, 553, 554, 555, 556, 557, 558, 559, 560, 561, 562, 563, 564, 565, 566, 567, 568, 569, 570, 571, 572, 573, 574, 575, 576, 577, 578, 579, 580, 581, 582, 583, 584, 585, 586, 587, 588, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599, 600, 601, 602, 603, 604, 605, 606, 607, 608, 609, 610, 611, 612, 613, 614, 615, 616, 617, 618, 619, 620, 621, 622, 623, 624, 625, 626, 627, 628, 629, 630, 631, 632, 633, 634, 635, 636, 637, 638, 639, 640, 641, 642, 643, 644, 645, 646, 647, 648, 649, 650, 651, 652, 653, 654, 655, 656, 657, 658, 6/home/l/dafoam/packages/miniconda3/lib/python3.8/site-packages/openmdao/core/component.py:140: OMDeprecationWarning:'cruise.coupling.solver' <class DAFoamSolver>: It appears this component mixes distributed/non-distributed inputs and outputs, so it may break starting with OpenMDAO 3.25, where the convention used when passing data between distributed and non-distributed inputs and outputs within a matrix free component will change. See https://github.com/OpenMDAO/POEMs/blob/master/POEM_075.md for details.
```

5. _and some residuals seem also weird:_
```
Printing Primal Residual Statistics.
U Residual Norm2: (103937.1369676273 12969.1428411269 5.54004279294636e-12)
U Residual Mean: (230.708942790629 18.66944257296205 2.716931902033349e-14)
U Residual Max: (33272.53893565353 5468.617226470547 1.222129631901349e-12)
p Residual Norm2: 127.3067535261034
p Residual Mean: 0.1709506084874724
p Residual Max: 81.65208747236603
nuTilda Residual Norm2: 29.61520970512968
nuTilda Residual Mean: 0.04749799128506413
nuTilda Residual Max: 16.95630771029906
phi Residual Norm2: 2.904830801753964
phi Residual Mean: 0.001618112317877623
phi Residual Max: 0.9002619739004367
End
```
 (Please find the whole logs here: 
[logOpt.txt](https://github.com/user-attachments/files/18082626/logOpt.txt)
[opt_IPOPT.txt](https://github.com/user-attachments/files/18082629/opt_IPOPT.txt))

**So my questions are:**

1. **Am I getting reasonable and reliable results? If not, how should I change in the runScript and how could I get better result?** 
2. **How could I get J0 in the Input Parameters? According to my understanding, I may calculate the difference of "UData" and U coming from a simulation with SA model(same to making betaFieldInversion=1 everywhere and simulating), then I may use the square of the norm2 of that difference as J0. Am I correct?**

Thanks for your help!

**Comment by SomeChilliness** [comment]
Hello all,

After checking my runScript, I realized that I did not correctly assign "min" and "max" here:
```
            "UData": {
                "type": "fieldInversion",
                "source": "boxToCell",
                "min": [59.44, -10.0, -10.0],
                "max": [79.44, 10.0, 10.0],
                "data": "UData",
                "scale": 1,
                "addToAdjoint": True,
                "weightedSum": True,
                "weight": 1 / J0,
            },
```

Thus I modified it according to my UData file, now the results are all good and seem to be reasonable!!!

I hope this could help ones who may observe the same warnings.

But still, I have not yet figured out if I were correct for getting J0 as my second question shows. 

Also, Primal Residual Statistics after running my new runScript, appeared to be large. I guess it could just be because of the case itself...

**Comment by friedenhe** [comment]
It is good to hear that it worked eventually. J0 is a scaling factor that balances the weight between the error and beta regulation term. Regarding the large residual, it should be OK, I believe it is partially caused by the large aspect ratio of your mesh, such that the flow did not converge very well.

**Reply by SomeChilliness** [reply]
Thanks a lot for answering my questions!

**Comment by 15175829310** [comment]
What is the function of "min" and "max"? Do they have to cover the entire flow field?

**Comment by 15175829310** [comment]
What is the function of "min" and "max"? Do they have to cover the entire flow field?

**Reply by SomeChilliness** [reply]
According to my case above, yes, I think. They may depend on your data file(Udata for example), the code works well as I set "min" and "max" to cover the range of flow field velocity variation in my Udata file.

**Reply by 15175829310** [reply]
Thank you for your reply. I have another question. Have you ever used surface pressure to perform flow field inversion? Does using surface pressure require full-field data?

**Reply by SomeChilliness** [reply]
I have not tried surface pressure for flow field inversion as I am focusing on velocity field in my on-going work. So I am not sure about your question, but I noticed that the patch name may have to be assigned in a test Script(line 98 in this [link](https://github.com/mdolab/dafoam/blob/main/tests/runTests_DASimpleFoamFieldInversionObjectiveFunctions.py) ), perheps entire filed data are not necessary for surface pressure... I think you may try and see if incomplete data would work and check if there are such tutorials using surface pressure and what was the surface pressure data like...

**Reply by friedenhe** [reply]
@15175829310 There is an option called `surface`. You need to put a pData file into the 0 folder and use fixedValue BC for the wall; then, you need to manually set the BC values for this pData file. Then, you should be good to go.

**Comment by 15175829310** [comment]
Thanks for your reply, it's very helpful. I am a new user and not very familiar with this, so I often encounter some problems. Why is it that the y+ calculated by openfoam is ≈30, but the y+ calculated by DArhoSimplecFoam is ≈0.016? I don't know what the problem is. I look forward to your reply.

**Reply by SomeChilliness** [reply]
Maybe you could check if you are using the same wallfuction or turbulence model... I am not very sure...
